File: src/server/services/program-matching/eligibility.ts
import { ASPIRATIONAL_REQUIREMENT_TYPES } from "@/lib/program-matching/config";
import type {
  EligibilityStatus,
  MatchingProfile,
  ProgramRiskFlag,
  RequirementEvaluation,
} from "@/lib/program-matching/types";
import {
  compareLanguageLevel,
  compareNumericRequirement,
  deadlineStatus,
  isUnknown,
} from "./compare";

const ASPIRATIONAL = new Set<string>(ASPIRATIONAL_REQUIREMENT_TYPES);

function isAspirationalRequirement(type: string) {
  return ASPIRATIONAL.has(type);
}

export type RequirementInput = {
  type: string;
  required: boolean;
  operator?: string | null;
  valueJson?: string | null;
  description?: string | null;
  hardExclusion?: boolean;
  sourceUrl?: string | null;
};

export type CycleInput = {
  applicationDeadline?: Date | string | null;
  applicantCategory?: string | null;
};

function parseValue(json: string | null | undefined): Record<string, unknown> {
  if (!json) return {};
  try {
    const v = JSON.parse(json);
    return v && typeof v === "object" ? v : {};
  } catch {
    return {};
  }
}

function languageFromProfile(profile: MatchingProfile, lang: string) {
  if (/english|en/i.test(lang)) return profile.englishLevel;
  if (/italian|it/i.test(lang)) return profile.italianLevel;
  return "UNKNOWN";
}

function cityEquals(a: string, b: string) {
  return a.trim().toLowerCase() === b.trim().toLowerCase();
}

/** Structural degree levels that must align for a hard match. */
export function degreesCompatible(
  desired: string,
  programDegreeLevel: string
): boolean {
  if (desired === programDegreeLevel) return true;
  // Legacy catalog rows may store foundation pathways as OTHER.
  if (desired === "FOUNDATION" && programDegreeLevel === "OTHER") return true;
  return false;
}

export function evaluateRequirement(
  profile: MatchingProfile,
  req: RequirementInput
): RequirementEvaluation {
  const value = parseValue(req.valueJson);
  const description = req.description || req.type;
  const base = {
    type: req.type as RequirementEvaluation["type"],
    description,
    required: req.required,
    hardExclusion: false,
    sourceUrl: req.sourceUrl,
  };

  switch (req.type) {
    case "LANGUAGE": {
      const level = String(value.level ?? value.minLevel ?? "");
      const lang = String(value.language ?? "English");
      const status = compareLanguageLevel(languageFromProfile(profile, lang), level);
      return { ...base, status };
    }
    case "SAT": {
      const score = Number(value.score ?? value.min ?? value.minimum);
      const status = compareNumericRequirement(
        profile.sat,
        req.operator || ">=",
        Number.isFinite(score) ? score : null
      );
      return { ...base, status };
    }
    case "TOLC": {
      const test = String(value.test ?? "TOLC");
      const score = Number(value.score ?? value.min);
      const studentScore = profile.tolc[test];
      const status = compareNumericRequirement(
        typeof studentScore === "number" ? studentScore : "UNKNOWN",
        req.operator || ">=",
        Number.isFinite(score) ? score : null
      );
      return { ...base, status };
    }
    case "ADMISSION_TEST":
    case "INTERVIEW":
    case "PORTFOLIO":
    case "CURRICULAR_CREDITS":
    case "SUBJECT_PREREQUISITE":
    case "ACADEMIC_GRADE": {
      return { ...base, status: "UNKNOWN" };
    }
    case "EDUCATION": {
      const requiredLevel = String(value.degreeLevel ?? value.level ?? "");
      if (!requiredLevel) return { ...base, status: "UNKNOWN" };
      if (isUnknown(profile.desiredDegreeLevel)) return { ...base, status: "UNKNOWN" };
      if (degreesCompatible(String(profile.desiredDegreeLevel), requiredLevel)) {
        return { ...base, status: "MET" };
      }
      return { ...base, status: "NOT_MET" };
    }
    case "CITIZENSHIP":
    case "RESIDENCY_STATUS": {
      if (profile.applicantCategory === "UNKNOWN") return { ...base, status: "UNKNOWN" };
      const allowed = Array.isArray(value.categories)
        ? value.categories.map(String)
        : [];
      if (allowed.length === 0) return { ...base, status: "UNKNOWN" };
      return {
        ...base,
        status: allowed.includes(profile.applicantCategory) ? "MET" : "NOT_MET",
      };
    }
    default:
      return { ...base, status: "UNKNOWN" };
  }
}

/**
 * Hard exclusion only for:
 * - structural degree level (bachelor / master / foundation / single-cycle)
 * - teaching-language preference
 * - avoid-list geography (preferred cities are fit-only / secondary)
 *
 * Certificates, admission tests, deadlines, citizenship, etc. are prep /
 * preference signals and must not block the match.
 */
export function evaluateEligibility(input: {
  profile: MatchingProfile;
  programDegreeLevel: string;
  teachingLanguages: string[];
  campusCity?: string | null;
  region?: string | null;
  requirements: RequirementInput[];
  cycles: CycleInput[];
  dataConfidence: string;
  usingPreviousYear?: boolean;
}): {
  status: EligibilityStatus;
  evaluations: RequirementEvaluation[];
  risks: ProgramRiskFlag[];
} {
  const { profile } = input;
  const evaluations = input.requirements.map((r) => evaluateRequirement(profile, r));
  const risks: ProgramRiskFlag[] = [];

  if (profile.applicantCategory === "UNKNOWN") {
    risks.push("APPLICANT_CATEGORY_UNVERIFIED", "CURATOR_REVIEW_REQUIRED");
  }

  const degreeMismatch =
    !isUnknown(profile.desiredDegreeLevel) &&
    profile.desiredDegreeLevel !== "OTHER" &&
    !degreesCompatible(String(profile.desiredDegreeLevel), input.programDegreeLevel);

  if (degreeMismatch) {
    evaluations.unshift({
      type: "EDUCATION",
      description: `Programme is ${input.programDegreeLevel}, student seeks ${profile.desiredDegreeLevel}`,
      status: "NOT_MET",
      required: true,
      hardExclusion: true,
    });
  }

  const city = (input.campusCity ?? "").trim();
  const region = (input.region ?? "").trim();

  if (city && profile.excludedCities.some((c) => cityEquals(c, city))) {
    evaluations.push({
      type: "OTHER",
      description: `City ${city} is in student's avoid list`,
      status: "NOT_MET",
      required: true,
      hardExclusion: true,
    });
  } else if (
    region &&
    profile.excludedRegions.some((r) => cityEquals(r, region))
  ) {
    evaluations.push({
      type: "OTHER",
      description: `Region ${region} is excluded by student preferences`,
      status: "NOT_MET",
      required: true,
      hardExclusion: true,
    });
  }

  // Teaching-language preference is a hard structural filter.
  const langPref = profile.preferredTeachingLanguages.map((l) => l.toLowerCase());
  if (langPref.length > 0 && input.teachingLanguages.length > 0) {
    const overlap = input.teachingLanguages.some((l) =>
      langPref.some(
        (p) =>
          l.toLowerCase().includes(p.toLowerCase()) ||
          p.toLowerCase().includes(l.toLowerCase())
      )
    );
    if (!overlap) {
      evaluations.push({
        type: "LANGUAGE",
        description: `Teaching language ${input.teachingLanguages.join(", ")} vs preference ${profile.preferredTeachingLanguages.join(", ")}`,
        status: "NOT_MET",
        required: true,
        hardExclusion: true,
      });
    }
  }

  const hardNotMet = evaluations.filter((e) => e.hardExclusion && e.status === "NOT_MET");

  // Deadlines are informational only — never hard-exclude.
  if (!input.usingPreviousYear) {
    const relevantDeadlines = input.cycles
      .filter((c) => {
        if (!c.applicantCategory || c.applicantCategory === "ALL") return true;
        if (profile.applicantCategory === "UNKNOWN") return true;
        return c.applicantCategory === profile.applicantCategory;
      })
      .map((c) => deadlineStatus(c.applicationDeadline));

    if (relevantDeadlines.some((d) => d === "SOON")) {
      risks.push("DEADLINE_SOON");
    } else if (
      relevantDeadlines.filter((d) => d !== "UNKNOWN").length > 0 &&
      relevantDeadlines.filter((d) => d !== "UNKNOWN").every((d) => d === "PASSED")
    ) {
      risks.push("CURATOR_REVIEW_REQUIRED");
    }
  } else {
    risks.push("USING_PREVIOUS_YEAR_DATA", "CURATOR_REVIEW_REQUIRED");
  }

  if (hardNotMet.length > 0) {
    return { status: "NOT_ELIGIBLE", evaluations, risks };
  }

  // Non-structural requirement gaps never flip to NOT_ELIGIBLE.
  const blockingUnknown = evaluations.filter(
    (e) =>
      e.required &&
      e.status === "UNKNOWN" &&
      !isAspirationalRequirement(e.type) &&
      e.type !== "CITIZENSHIP" &&
      e.type !== "RESIDENCY_STATUS" &&
      e.type !== "OTHER"
  );

  if (
    blockingUnknown.length > 0 ||
    input.dataConfidence !== "HIGH" ||
    profile.applicantCategory === "UNKNOWN" ||
    input.usingPreviousYear
  ) {
    if (
      blockingUnknown.length === 0 &&
      profile.applicantCategory !== "UNKNOWN" &&
      !input.usingPreviousYear &&
      input.dataConfidence !== "HIGH"
    ) {
      return { status: "LIKELY_ELIGIBLE", evaluations, risks };
    }
    if (
      profile.applicantCategory === "UNKNOWN" ||
      input.usingPreviousYear ||
      blockingUnknown.length > 0
    ) {
      return { status: "NEEDS_REVIEW", evaluations, risks };
    }
    return { status: "LIKELY_ELIGIBLE", evaluations, risks };
  }

  return { status: "ELIGIBLE", evaluations, risks };
}

File: backend/app/services/matching/eligibility.py
from app.schemas.matching import (
    EligibilityRequest,
    EligibilityResponse,
    RequirementEvaluation,
    RequirementInput,
)
from app.services.matching.compare import (
    UNKNOWN,
    compare_language_level,
    compare_numeric_requirement,
    deadline_status,
    is_unknown,
)

ASPIRATIONAL_TYPES = {
    "LANGUAGE",
    "SAT",
    "TOLC",
    "ADMISSION_TEST",
    "INTERVIEW",
    "PORTFOLIO",
    "CURRICULAR_CREDITS",
    "SUBJECT_PREREQUISITE",
    "ACADEMIC_GRADE",
}


def degrees_compatible(desired: str, program: str) -> bool:
    return desired == program or (desired == "FOUNDATION" and program == "OTHER")


def _language_from_profile(request: EligibilityRequest, language: str) -> str:
    if language.lower() in {"english", "en"}:
        return str(request.profile.english_level)
    if language.lower() in {"italian", "it"}:
        return str(request.profile.italian_level)
    return UNKNOWN


def evaluate_requirement(
    request: EligibilityRequest, requirement: RequirementInput
) -> RequirementEvaluation:
    description = requirement.description or requirement.type
    base = {
        "type": requirement.type,
        "description": description,
        "required": requirement.required,
        "source_url": requirement.source_url,
    }
    value = requirement.value
    match requirement.type:
        case "LANGUAGE":
            level = str(value.get("level") or value.get("minLevel") or "")
            language = str(value.get("language") or "English")
            status = compare_language_level(_language_from_profile(request, language), level)
        case "SAT":
            minimum = value.get("score", value.get("min", value.get("minimum")))
            required_score = float(minimum) if isinstance(minimum, (int, float)) else None
            status = compare_numeric_requirement(
                request.profile.sat, requirement.operator, required_score
            )
        case "TOLC":
            test = str(value.get("test") or "TOLC")
            minimum = value.get("score", value.get("min"))
            required_score = float(minimum) if isinstance(minimum, (int, float)) else None
            status = compare_numeric_requirement(
                request.profile.tolc.get(test, UNKNOWN), requirement.operator, required_score
            )
        case "EDUCATION":
            required_level = str(value.get("degreeLevel") or value.get("level") or "")
            if not required_level or is_unknown(request.profile.desired_degree_level):
                status = "UNKNOWN"
            else:
                status = (
                    "MET"
                    if degrees_compatible(request.profile.desired_degree_level, required_level)
                    else "NOT_MET"
                )
        case "CITIZENSHIP" | "RESIDENCY_STATUS":
            categories = [str(item) for item in value.get("categories", [])]
            if request.profile.applicant_category == "UNKNOWN" or not categories:
                status = "UNKNOWN"
            else:
                status = "MET" if request.profile.applicant_category in categories else "NOT_MET"
        case _:
            status = "UNKNOWN"
    return RequirementEvaluation(**base, status=status)


def _same_place(first: str, second: str) -> bool:
    return first.strip().lower() == second.strip().lower()


def evaluate_eligibility(request: EligibilityRequest) -> EligibilityResponse:
    evaluations = [
        evaluate_requirement(request, requirement) for requirement in request.requirements
    ]
    risks: list[str] = []
    profile = request.profile

    if profile.applicant_category == "UNKNOWN":
        risks.extend(["APPLICANT_CATEGORY_UNVERIFIED", "CURATOR_REVIEW_REQUIRED"])

    if not is_unknown(profile.desired_degree_level) and profile.desired_degree_level != "OTHER":
        if not degrees_compatible(profile.desired_degree_level, request.program_degree_level):
            evaluations.insert(
                0,
                RequirementEvaluation(
                    type="EDUCATION",
                    description=(
                        f"Programme is {request.program_degree_level}, student seeks "
                        f"{profile.desired_degree_level}"
                    ),
                    status="NOT_MET",
                    required=True,
                    hard_exclusion=True,
                ),
            )

    if request.campus_city and any(
        _same_place(city, request.campus_city) for city in profile.excluded_cities
    ):
        evaluations.append(
            RequirementEvaluation(
                type="OTHER",
                description=f"City {request.campus_city} is in student's avoid list",
                status="NOT_MET",
                required=True,
                hard_exclusion=True,
            )
        )
    elif request.region and any(
        _same_place(region, request.region) for region in profile.excluded_regions
    ):
        evaluations.append(
            RequirementEvaluation(
                type="OTHER",
                description=f"Region {request.region} is excluded by student preferences",
                status="NOT_MET",
                required=True,
                hard_exclusion=True,
            )
        )

    preferences = [language.lower() for language in profile.preferred_teaching_languages]
    if preferences and request.teaching_languages:
        has_overlap = any(
            preference in language.lower() or language.lower() in preference
            for language in request.teaching_languages
            for preference in preferences
        )
        if not has_overlap:
            evaluations.append(
                RequirementEvaluation(
                    type="LANGUAGE",
                    description=(
                        f"Teaching language {', '.join(request.teaching_languages)} vs preference "
                        f"{', '.join(profile.preferred_teaching_languages)}"
                    ),
                    status="NOT_MET",
                    required=True,
                    hard_exclusion=True,
                )
            )

    if request.using_previous_year:
        risks.extend(["USING_PREVIOUS_YEAR_DATA", "CURATOR_REVIEW_REQUIRED"])
    else:
        deadline_states = [
            deadline_status(cycle.application_deadline)
            for cycle in request.cycles
            if not cycle.applicant_category
            or cycle.applicant_category == "ALL"
            or profile.applicant_category == "UNKNOWN"
            or cycle.applicant_category == profile.applicant_category
        ]
        if "SOON" in deadline_states:
            risks.append("DEADLINE_SOON")
        elif [item for item in deadline_states if item != "UNKNOWN"] and all(
            item == "PASSED" for item in deadline_states if item != "UNKNOWN"
        ):
            risks.append("CURATOR_REVIEW_REQUIRED")

    if any(item.hard_exclusion and item.status == "NOT_MET" for item in evaluations):
        return EligibilityResponse(status="NOT_ELIGIBLE", evaluations=evaluations, risks=risks)

    blocking_unknown = [
        item
        for item in evaluations
        if item.required
        and item.status == "UNKNOWN"
        and item.type not in ASPIRATIONAL_TYPES | {"CITIZENSHIP", "RESIDENCY_STATUS", "OTHER"}
    ]
    needs_review = bool(
        blocking_unknown
        or profile.applicant_category == "UNKNOWN"
        or request.using_previous_year
    )
    if needs_review or request.data_confidence != "HIGH":
        status = "NEEDS_REVIEW" if needs_review else "LIKELY_ELIGIBLE"
        return EligibilityResponse(status=status, evaluations=evaluations, risks=risks)
    return EligibilityResponse(status="ELIGIBLE", evaluations=evaluations, risks=risks)
